node/src/components/consensus/highway_core/active_validator.rs

```diff
@@ -478,82 +483,268 @@ mod tests {
+            (effects, proposal_wunit)
+        }
+
+        /// Handle new unit by validator `vidx`.
+        /// Since all validator use the same state, that unit should be added already. Panics if
+        /// not. Returns effect created when handling new unit.
+        fn handle_new_unit(
+            &mut self,
+            vidx: ValidatorIndex,
+            uhash: &<TestContext as Context>::Hash,
+        ) -> Vec<Effect<TestContext>> {
+            let validator = self.validators.get_mut(&vidx).unwrap();
+            let delivery_timestamp = self.state.unit(uhash).timestamp + 1.into();
+            let effects = validator.on_new_unit(
+                uhash,
+                delivery_timestamp,
+                &self.state,
+                self.instance_id,
+                &mut self.rng,
+            );
+            self.schedule_timer(vidx, &effects);
+            self.add_new_unit(&effects);
+            effects
+        }
+
+        /// Schedules new timers, if any was returned as an effect.
+        fn schedule_timer(&mut self, vidx: ValidatorIndex, effects: &[Effect<TestContext>]) {
+            let new_timestamps: Vec<Timestamp> = effects
+                .iter()
+                .filter_map(|eff| {
+                    if let Effect::ScheduleTimer(timestamp) = eff {
+                        Some(*timestamp)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+            match *new_timestamps {
+                [] => (),
+                [timestamp] => {
+                    let _ = self.timers.insert((timestamp, vidx));
+                }
+                _ => panic!(
+                    "Expected at most one timer to be scheduled: {:?}",
+                    new_timestamps
+                ),
+            }
+        }
+
+        /// Adds new unit, if any, to the state.
+        fn add_new_unit(&mut self, effects: &[Effect<TestContext>]) {
+            let new_units: Vec<_> = effects
+                .iter()
+                .filter_map(|eff| {
+                    if let Effect::NewVertex(ValidVertex(Vertex::Unit(swunit))) = eff {
+                        Some(swunit)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+            match *new_units {
+                [] => (),
+                [unit] => {
+                    let _ = self.state.add_unit(unit.clone()).unwrap();
+                }
+                _ => panic!(
+                    "Expected at most one timer to be scheduled: {:?}",
+                    new_units
+                ),
+            }
+        }
+
+        /// Returns hash of the newly finalized unit.
+        fn next_finalized(&mut self) -> Option<&<TestContext as Context>::Hash> {
+            self.fd.next_finalized(&self.state)
+        }
+    }
+
+    fn unwrap_single<T: Debug + Clone>(vec: &[T]) -> T {
```

Comment:
There's alsoready an [`Itertools::exactly_one`](https://docs.rs/itertools/0.9.0/itertools/trait.Itertools.html#method.exactly_one); but I guess it's not _exactly_ equivalent to this.